### PR TITLE
Use map_rows instead of list of indices

### PR DESCRIPTION
**Pull request by blegat**
Broadcasting on a list of indices would not be very efficient because of type instability. Moreover, we had the issue that the slack and dual vector were a vector of vector instead of a vector. This PR fixes those issues using the `MatrixOptInterface.rows` function and function barrier technique on constraint types as it is common in MOI.

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -70,17 +70,14 @@ end
 mutable struct Optimizer{OT <: MOI.ModelLike} <: MOI.AbstractOptimizer
     optimizer::OT
     primal_optimal::Vector{Float64}  # solution
-    dual_optimal::Vector{Union{Vector{Float64}, Float64}}  # refer - https://github.com/jump-dev/DiffOpt.jl/issues/21#issuecomment-651130485
     var_idx::Vector{VI}
-    con_idx::Vector{CI}
 
     function Optimizer(optimizer_constructor::OT) where {OT <: MOI.ModelLike}
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:75]: don't you have one excess `zeros(0)` below?

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -472,21 +449,54 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
 end
 
 """
-    π(cones::Array{<: MOI.AbstractVectorSet}, v::Vector{Vector{Float64}})
+    π(v::Vector{Float64}, model, conic_form, index_map)
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:452]: this will need documenting what conic_form and index_map need to be

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -472,21 +449,54 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
...
+    end
+end
 
 
 """
-    Dπ(cones::Array{<: MOI.AbstractVectorSet}, v::Vector{Vector{Float64}})
+    Dπ(v::Vector{Float64}, model, conic_form, index_map)
 
 Find gradient of projection of vectors in `v` on product of `cones`.
 For more info, refer https://github.com/matbesancon/MathOptSetDistances.jl
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:472]:
```suggestion
For more info, refer to https://github.com/matbesancon/MathOptSetDistances.jl
```

**Comment by matbesancon** [comment]
overall this will need a bit of documentation

File: src/MOI_wrapper.jl [hunk 4]
```diff
@@ -472,24 +448,90 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
 end
 
 """
-    π(cones::Array{<: MOI.AbstractVectorSet}, v::Vector{Vector{Float64}})
+    π(v::Vector{Float64}, model, conic_form, index_map)
 
-Find projection of vectors in `v` on product of `cones`.
-For more info, refer https://github.com/matbesancon/MathOptSetDistances.jl
+Given a `model`, its `conic_form` and the `index_map` from the indices of
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:453]: Should maybe say what type / behaviour is expected for conic_form and index_map

File: src/MOI_wrapper.jl [hunk 5]
```diff
@@ -472,24 +448,90 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
...
+end
 
 
 """
-    Dπ(cones::Array{<: MOI.AbstractVectorSet}, v::Vector{Vector{Float64}})
+    Dπ(v::Vector{Float64}, model, conic_form, index_map)
 
-Find gradient of projection of vectors in `v` on product of `cones`.
-For more info, refer https://github.com/matbesancon/MathOptSetDistances.jl
+Given a `model`, its `conic_form` and the `index_map` from the indices of
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:473]: same as for the projection

File: src/MOI_wrapper.jl [hunk 6]
```diff
@@ -472,24 +448,90 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
...
+For more info, refer to https://github.com/matbesancon/MathOptSetDistances.jl
 """
-Dπ(cones, v) = MOSD.projection_gradient_on_set(MOSD.DefaultDistance(), v, cones)
+function Dπ(v::Vector{T}, model, conic_form, index_map) where T
+    return BlockDiagonals.BlockDiagonal(
+        map_rows(model, conic_form, index_map, Nested{Matrix{T}}()) do ci, r
+            MOSD.projection_gradient_on_set(
+                MOSD.DefaultDistance(),
+                v[r],
+                MOI.dual_set(MOI.get(model, MOI.ConstraintSet(), ci))
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:485]:
```suggestion
                MOI.dual_set(MOI.get(model, MOI.ConstraintSet(), ci)),
```

**Comment by matbesancon** [comment]
the changes look good. Since this is fairly non-trivial, it could be good to add something in the docs?

**Comment by matbesancon** [comment]
in any case this will be merged before #61 which still needs an update

**Comment by blegat** [comment]
Not sure what to add in the docs, this is mostly internal and the docstring of `map_rows` is added automatically

File: src/MOI_wrapper.jl [hunk 7]
```diff
@@ -472,24 +448,93 @@ function MOI.modify(model::Optimizer, ci::CI{F, S}, chg::MOI.AbstractFunctionMod
...
+function _map_rows!(f::Function, x::Vector, model, conic_form::MatOI.GeometricConicForm, index_map::MOIU.DoubleDicts.IndexWithType{F, S}, map_mode, k) where {F, S}
+    for ci in MOI.get(model, MOI.ListOfConstraintIndices{F, S}())
+        r = MatOI.rows(conic_form, index_map[ci])
+        k += 1
+        _assign_mapped!(x, f(ci, r), r, k, map_mode)
+    end
+    return k
+end
+
+_map_output(conic_form, ::Nested{T}) where {T} = Vector{T}(undef, length(conic_form.dimension))
```
Comment (matbesancon) [inline src/MOI_wrapper.jl:510]: this function name does not reflect what it does, "initialize_x" would be better? Not sure what x could be here

Comment (blegat) [reply src/MOI_wrapper.jl]: Done in latest commit

**Pull request merged**